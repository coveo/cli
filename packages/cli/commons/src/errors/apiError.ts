import {red} from 'chalk';
import dedent from 'ts-dedent';
import {PrintableError, SeverityLevel} from './printableError';

export interface APIErrorResponse {
  message: string;
  errorCode: string;
  requestID: string;
}

export interface AxiosErrorFromAPI {
  response: {
    status: number;
    data: APIErrorResponse;
  };
}

export const APIErrorSchema = {
  type: 'object',
  required: ['errorCode', 'message'],
  properties: {
    errorCode: {
      type: 'string',
    },
    message: {
      type: 'string',
    },
  },
};

export const AxiosErrorFromAPISchema = {
  type: 'object',
  required: ['response'],
  properties: {
    response: {
      type: 'object',
      required: ['status', 'data'],
      properties: {
        status: {
          type: 'integer',
        },
        data: APIErrorSchema,
      },
    },
  },
};

export class APIError extends PrintableError {
  public constructor(
    error: APIErrorResponse | AxiosErrorFromAPI,
    tagLine?: string
  ) {
    super(SeverityLevel.Error);
    let errorCode: string;
    let message: string;
    let status: number | undefined;
    if (this.isFromAxios(error)) {
      errorCode = error.response.data.errorCode;
      message = error.response.data.message;
      status = error.response.status;
    } else {
      errorCode = error.errorCode;
      message = error.message;
    }

    this.name = `APIError - ${errorCode}`;
    this.message = dedent`
    ${tagLine}
    ${status ? `Status code: ${status}\n` : ''}
    Error code: ${red(errorCode)}
    Message: ${red(message)}
    `;
  }

  private isFromAxios(
    error: APIErrorResponse | AxiosErrorFromAPI
  ): error is AxiosErrorFromAPI {
    return 'response' in error;
  }
}
