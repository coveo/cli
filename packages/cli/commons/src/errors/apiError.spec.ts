import {CoveoPlatformClientError} from '@coveo/platform-client';
import {APIError} from './apiError';

describe('APIError', () => {
  it('should print the tagline', () => {
    expect(() => {
      throw new APIError(new CoveoPlatformClientError(), 'This is a tagline');
    }).toThrowErrorMatchingSnapshot();
  });

  it('should not print a tagline', () => {
    expect(() => {
      throw new APIError(new CoveoPlatformClientError());
    }).toThrowErrorMatchingSnapshot();
  });

  describe('when the error is of type APIErrorResponse', () => {
    it.each([
      {
        title: 'it should print the details',
        response: () => {
          const err = new CoveoPlatformClientError();
          err.detail = 'something went wrong';
          return err;
        },
      },
      {
        title: 'it should print the request Id',
        response: () => {
          const err = new CoveoPlatformClientError();
          err.xRequestId = '123456qwerty';
          return err;
        },
      },
      {
        title: 'it should print the error code',
        response: () => {
          const err = new CoveoPlatformClientError();
          err.title = 'YOU_SHALL_NOT_GET_MODE_INFO';
          return err;
        },
      },
      {
        title: 'it should print everything',
        response: () => {
          const err = new CoveoPlatformClientError();
          err.xRequestId = '123456qwerty';
          err.title = 'YOU_SHALL_NOT_GET_MODE_INFO';
          err.detail = 'something went wrong';
          return err;
        },
      },
    ])(`$title`, ({response}) => {
      expect(() => {
        throw new APIError(response(), 'This is a tagline');
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
