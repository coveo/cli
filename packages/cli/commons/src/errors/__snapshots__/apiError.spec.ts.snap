// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APIError should not print a tagline 1`] = `
"
Status code: undefined
Error code: [31mundefined[39m
Message: [31mundefined[39m
Request ID: [31mundefined[39m"
`;

exports[`APIError should print the tagline 1`] = `
"
This is a tagline
Status code: undefined
Error code: [31mundefined[39m
Message: [31mundefined[39m
Request ID: [31mundefined[39m"
`;

exports[`APIError when the error is of type APIErrorResponse it should print everything 1`] = `
"
This is a tagline
Status code: undefined
Error code: [31mYOU_SHALL_NOT_GET_MODE_INFO[39m
Message: [31msomething went wrong[39m
Request ID: [31m123456qwerty[39m"
`;

exports[`APIError when the error is of type APIErrorResponse it should print the details 1`] = `
"
This is a tagline
Status code: undefined
Error code: [31mundefined[39m
Message: [31msomething went wrong[39m
Request ID: [31mundefined[39m"
`;

exports[`APIError when the error is of type APIErrorResponse it should print the error code 1`] = `
"
This is a tagline
Status code: undefined
Error code: [31mYOU_SHALL_NOT_GET_MODE_INFO[39m
Message: [31mundefined[39m
Request ID: [31mundefined[39m"
`;

exports[`APIError when the error is of type APIErrorResponse it should print the request Id 1`] = `
"
This is a tagline
Status code: undefined
Error code: [31mundefined[39m
Message: [31mundefined[39m
Request ID: [31m123456qwerty[39m"
`;
